### DISCUSSION TITLE
IncidentStrategy and retry an Activity

### DISCUSSION THREAD

**Discussion by kimseb** [elsa-workflows/elsa-core#6131] [Q&A]
I want to implement incident handling for activities that fails and be able to restart the activities from: 
- my IncidentHandler (with the same data again) 
- and/or Elsa Studio under "Instances" (with the same data again) 

Is that possible? 

I have not found any example project with IncidentHandling/IncidentStrategies and I can't seem to find code example for retrying/re-running an activity if it fails in a workflow.

I would love if this could be added somewhere, in wiki or a sample-project. :) And I need help with this....

**Comment by kimseb** [comment]
I wonder if something like this would be possible or if there is any way to make this work?/Any alternative way of doing it?

**The IncidentStrategy**
```
        public async void HandleIncident(ActivityExecutionContext activityContext)
        {
            WorkflowExecutionContext workflowContext = activityContext.WorkflowExecutionContext;

            //Email or log

            long numberOfExecusions = activityContext.ExecutionCount;
            switch (numberOfExecusions)
            {
                case 0:
                    await RunActivityAgainWithDelay(1, workflowContext, activityContext);
                    break;
                case 1:
                    await RunActivityAgainWithDelay(2, workflowContext, activityContext);
                    break;
                case 2:
                    await RunActivityAgainWithDelay(5, workflowContext, activityContext);
                    break;
                case 3:
                    await RunActivityAgainWithDelay(10, workflowContext, activityContext);
                    break;
                default:
                    //Email or log
                    break;
            }
        }
```

**The code that will schedule and (hopefully) re-run the activity that failed:**
```
private static async Task RunActivityAgainWithDelay(int minutes, WorkflowExecutionContext workflowContext, ActivityExecutionContext activityContext)
{
    await Task.Delay(TimeSpan.FromMinutes(minutes));
    var scheduleWorkOptions = new ScheduleWorkOptions()
    {
        CompletionCallback = OnCompleted, //To send email or log when its completed with a success
        ExistingActivityExecutionContext = activityContext,
    };

    var activityNode = activityContext.ActivityNode;
    workflowContext.Schedule(activityNode, activityContext, scheduleWorkOptions);
}
```

For now, it just stops after being Scheduled and gets status _Suspended_ in Elsa Studio under "Instances".

**Reply by kimseb** [reply]
> 
> **The code that will schedule and (hopefully) re-run the activity that failed:**
> 
> ```
> private static async Task RunActivityAgainWithDelay(int minutes, WorkflowExecutionContext workflowContext, ActivityExecutionContext activityContext)
> {
>     await Task.Delay(TimeSpan.FromMinutes(minutes));
>     var scheduleWorkOptions = new ScheduleWorkOptions()
>     {
>         CompletionCallback = OnCompleted, //To send email or log when its completed with a success
>         ExistingActivityExecutionContext = activityContext,
>     };
> 
>     var activityNode = activityContext.ActivityNode;
>     workflowContext.Schedule(activityNode, activityContext, scheduleWorkOptions);
> }
> ```
> 
> For now, it just stops after being Scheduled and gets status _Suspended_ in Elsa Studio under "Instances".

My colleague managed to start the scheduled activity by adding:
```
await _workflowDispatcher.DispatchAsync(new DispatchWorkflowInstanceRequest(workflowContext.Id), CancellationToken.None);
```
Beneath workflowContext.Schedule(...);

But ExecutionCount inside the ActivityContext is not incremented when doing this unfortunately. Our workaround is to count the number of Incidents inside the workflowContext instead, but it is not as viable.

**The next objective is** to start a failed activity manually. And right now we are attempting to do it by calling an endpoint and passing the workflow instance id as an url parameter. Then we use the same Dispatch-method but nothing happens. Perhaps we need to schedule it aswell like we do in the incidenthandler? Maybe a WorkflowInstanceStore is something we can use to get the workflow's context?

**Reply by kimseb** [reply]
Ok, restarting a workflow from a certain Activity (Status=Canceled for now) from an HTTP endpoint that tries to invoke it was not that easy :) If anyone knows how to do it please help.

**Reply by kimseb** [reply]
**Update: I have been able to fetch an activityExecutionContextState. Inside it resides the variables used in the activity but to reuse the scheduler and dispatcher we will need the activityExecutionContext and right now I do not know a way to recieve it...**

How do I get a hold of the ActivityExecutionContext? Is it possible?

```
app.MapGet("/workflow/{workflowId}", async ([FromRoute] string workflowId) =>
{
    var store = app.Services.GetRequiredService<IWorkflowInstanceStore>();

    WorkflowInstance oldWorkflow = await store.FindAsync(workflowId);
    WorkflowState state = oldWorkflow.WorkflowState;
    var activityContextState = state.ActivityExecutionContexts.First();

    var scheduleWorkOptions = new ScheduleWorkOptions()
    {
        //CompletionCallback = OnCompleted,
        //ExistingActivityExecutionContext = activityContext,
        //Input = activityContext.ActivityInput,
    };
    
    //workflowContext.Schedule(activityNode, activityContext, scheduleWorkOptions);

    var _workflowDispatcher = app.Services.GetRequiredService<IWorkflowDispatcher>();
    //await _workflowDispatcher.DispatchAsync(new DispatchWorkflowInstanceRequest(oldWorkflow.Id), CancellationToken.None);
});
```

**Comment by kimseb** [comment]
![{61DDAFA4-7265-460C-A3D5-B662BE630169}](https://github.com/user-attachments/assets/afa80a13-dc78-4021-9847-459a0e6cc922)
The greyed out "Retry" from the drop-menu, will it restart the whole workflow or just the failed activity? Why is it greyed out?

**Comment by yinzara** [comment]
Does anyone have any idea how to Retry a workflow in Elsa or why that option is grayed out or how to make it work again?

**Comment by sfmskywalker** [comment]
Current Elsa releases provide a supported retry path, so you no longer need to retain an `ActivityExecutionContext` and schedule it manually from an incident strategy.

Use the Alterations retry endpoint, `POST /alterations/workflows/retry`, and target the workflow instance plus the failed activity ID(s). It creates `ScheduleActivity` alterations, resets an existing faulted activity to `Running`, schedules it, and dispatches the workflow instance. The endpoint can also select all faulted activity IDs when none are supplied.

Take care with side effects: retried activities should be idempotent, or guarded/compensated appropriately. `ContinueWithIncidentsStrategy` remains useful when the desired behavior is to keep a workflow running with incidents, but it is no longer required just to obtain retry capability.

Relevant source: https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Alterations/Endpoints/Workflows/Retry/Endpoint.cs and https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Alterations/AlterationHandlers/ScheduleActivityHandler.cs. Related discussion: https://github.com/elsa-workflows/elsa-core/discussions/6215
